src/radical/entk/utils/provenance.py

```diff
@@ -1,36 +1,120 @@
 
-__copyright__ = 'Copyright 2023, The RADICAL-Cybertools Team'
+__copyright__ = 'Copyright 2024, The RADICAL-Cybertools Team'
 __license__   = 'MIT'
 
-from typing import Optional, Dict, List
+import glob
+
+from typing import Optional, Dict, List, Union
 
 import radical.utils as ru
 
-from .. import Pipeline
+from .. import Pipeline, Task
+
+_darshan_env = None
+
+
+# ------------------------------------------------------------------------------
+#
+def enable_darshan(pipelines: List[Pipeline],
+                   darshan_runtime_root: Optional[str] = None,
+                   modules: Optional[List[str]] = None) -> None:
+
+    if darshan_runtime_root:
+        if not darshan_runtime_root.startswith('/'):
+            raise RuntimeError('Path for the darshan installation '
+                               'should be an absolute path '
+                               f'(provided path: {darshan_runtime_root})')
+    else:
+        darshan_runtime_root = '$DARSHAN_RUNTIME_ROOT'
+
+    global _darshan_env
+
+    darshan_activation_cmds = []
+    for module in modules or []:
+        darshan_activation_cmds.append(f'module load {module}')
+        _darshan_env = ru.env_prep(pre_exec_cached=darshan_activation_cmds)
+
+    for pipeline in pipelines:
+        for stage in pipeline.stages:
+            for task in stage.tasks:
+
+                darshan_log_dir = '${RP_TASK_SANDBOX}/${RP_TASK_ID}_darshan'
+                darshan_enable  = (f'LD_PRELOAD="{darshan_runtime_root}'
+                                   '/lib/libdarshan.so" ')
+
+                if task.cpu_reqs.cpu_processes == 1:
+                    darshan_enable += 'DARSHAN_ENABLE_NONMPI=1 '
+
+                task.executable  = darshan_enable + task.executable
+                task.pre_launch += [f'mkdir -p {darshan_log_dir}']
+                task.pre_exec.extend(
+                    darshan_activation_cmds +
+                    [f'export DARSHAN_LOG_DIR_PATH={darshan_log_dir}'])
+
+
+# ------------------------------------------------------------------------------
+#
+def get_parsed_data(log: str, target_counters: Union[str, List[str]]) -> set:
+
+    global _darshan_env
+    data = set()
+
+    grep_patterns = '-e ' + ' -e '.join(ru.as_list(target_counters))
```

Comment:
it will result `-e pattern_1 -e pattern_2 ...`, since `join` method is applied to a string, which will be in between of the list elements (e.g., `'pattern_1' + ' -e ' + 'pattern_2'`), and then we add `-e` for the 1st element

Reply:
Ah, of course - stupid me! :-P